Sources/BSON/BSON.swift

```diff
@@ -77,6 +77,42 @@ public enum BSON {
         }
     }
 
+    /// Initialize a `BSON` from ExtendedJSON
+    /// Parameters:
+    ///   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for any `BSONValue`.
+    ///   - `keyPath`: an array of `Strings`s containing the enclosing JSON keys of the current json being passed in.
+    ///              This is used for error messages.
+    ///
+    /// Throws:
+    ///   - `InternalError`
+    internal init(fromExtJSON json: JSON, keyPath: [String]) throws {
+        // Spec requires that we try int32, try int64, then try double for decoding numbers
+        if let int32 = try Int32(fromExtJSON: json, keyPath: keyPath) {
+            self = int32.bson
+            return
+        }
+        if let int64 = try Int64(fromExtJSON: json, keyPath: keyPath) {
+            self = int64.bson
+            return
+        }
+        for bsonType in BSON.allBSONTypes.values {
+            if bsonType == Int32.self || bsonType == Int64.self || bsonType == BSONDocument.self {
+                continue
+            }
+            if let value = try bsonType.init(fromExtJSON: json, keyPath: keyPath) {
+                self = value.bson
+                return
+            }
+        }
+        // Document accepts any JSON object so it should be tried after all the more specific BSON types are tried
+        if let doc = try BSONDocument(fromExtJSON: json, keyPath: keyPath) {
```

Comment:
since this is something that should happen, can we invert this to be a `guard` instead?

Reply:
yep